mapper/collectd_mapper/src/batcher.rs

```diff
@@ -1,171 +1,199 @@
-use clock::{Clock, Timestamp};
+use crate::collectd::{CollectdMessage, OwnedCollectdMessage};
+use crate::error::*;
+use clock::Clock;
+use message_algos::{
+    Envelope, FilterDecision, GroupingPolicy, LinearDedupFilter, MessageBatcher as Batcher,
+    MessageFilter, MessageGroup, MessageGrouper, RetirementDecision, RetirementPolicy, Timestamp,
+};
 use mqtt_client::{Message, MqttClient, MqttMessageStream, Topic, TopicFilter};
 use std::sync::Arc;
 use thin_edge_json::{
     group::MeasurementGrouper, measurement::FlatMeasurementVisitor,
     serialize::ThinEdgeJsonSerializer,
 };
-use tokio::{
-    select,
-    sync::mpsc::{UnboundedReceiver, UnboundedSender},
-    time,
-    time::Duration,
-};
-use tracing::{error, log::warn};
-
-use crate::collectd::CollectdMessage;
-use crate::error::*;
-
-#[derive(Debug)]
-pub struct MessageBatch {
-    message_grouper: MeasurementGrouper,
+use tokio::sync::mpsc::{UnboundedReceiver, UnboundedSender};
+use tracing::{error, log::info, log::warn};
+
+/// We start a new batch upon receiving a message whose timestamp is farther away to the
+/// timestamp of first message in the batch than `delta` seconds.
+///
+/// Delta is inclusive.
+struct CollectdTimestampDeltaGroupingPolicy {
+    delta: f64,
 }
 
-impl MessageBatch {
-    fn start_batch(
-        collectd_message: CollectdMessage,
-        timestamp: Timestamp,
-    ) -> Result<Self, DeviceMonitorError> {
-        let mut message_grouper = MeasurementGrouper::new();
-        message_grouper.timestamp(&timestamp)?;
-
-        let mut message_batch = Self { message_grouper };
-
-        message_batch.add_to_batch(collectd_message)?;
+impl GroupingPolicy for CollectdTimestampDeltaGroupingPolicy {
+    type Message = OwnedCollectdMessage;
 
-        Ok(message_batch)
+    fn belongs_to_group(
+        &self,
+        message: &Envelope<Self::Message>,
+        group: &MessageGroup<Self::Message>,
+    ) -> bool {
+        let delta = group.first().message.timestamp() - message.message.timestamp();
```

Comment:
Need to factor in the `max message delay` as well, as the batching windows don't start from the timestamp of the first message but rather from `group.first().message.timestamp() - max.message.delay` as per the new batching specification.